File: backend/application/approve_leave_usecase.py
from backend.infrastructure.leave_repository import LeaveRepository
from backend.domain.leave_service import LeaveDomainService
from backend.domain.leave_entity import LeaveStatus

class ApproveLeaveUseCase:
    # Cấp 2: Chứa orchestration logic, phụ thuộc Repository & DomainService (SOLID - DI)
    def __init__(self, repo: LeaveRepository, domain_service: LeaveDomainService):
        self.repo = repo
        self.domain_service = domain_service

    def execute(self, leave_request_id: int) -> dict:
        leave_request = self.repo.get_leave_request(leave_request_id)
        if not leave_request:
            return {"success": False, "message": "Không tìm thấy đơn."}
        
        if leave_request.status != LeaveStatus.PENDING:
            return {"success": False, "message": "Đơn đã được xử lý."}

        employee = self.repo.get_employee(leave_request.employee_id)
        if not employee:
            return {"success": False, "message": "Không tìm thấy nhân viên."}

        # Gọi Domain Service để kiểm tra business rules
        if not self.domain_service.can_approve_leave(employee, leave_request.days):
            return {"success": False, "message": "Nhân viên không đủ ngày nghỉ phép."}

        # Orchestrate cập nhật repository
        self.repo.update_leave_status(leave_request.id, LeaveStatus.APPROVED)
        self.repo.update_employee_remaining_days(employee.id, leave_request.days)

        return {"success": True, "message": "Duyệt thành công!"}

File: backend/application/use_cases/ApproveLeaveUseCase.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { LeaveRepository } from '../../infrastructure/repositories/LeaveRepository.ts';
import { LeaveDomainService } from '../../domain/services/LeaveDomainService.ts';
import { LeaveStatus } from '../../domain/entities/LeaveEntity.ts';

export class ApproveLeaveUseCase {
  constructor(
    private leaveRepo: LeaveRepository,
    private leaveDomainService: LeaveDomainService
  ) {}

  /**
   * Cấp 2: Điều phối logic (Orchestration)
   */
  async execute(leaveRequestId: string): Promise<{ success: boolean; message: string }> {
    // 1. Lấy thông tin đơn (Infrastructure Level)
    const leaveRequest = await this.leaveRepo.getLeaveRequestById(leaveRequestId);
    if (!leaveRequest) {
      return { success: false, message: 'Leave request not found' };
    }

    if (leaveRequest.status !== LeaveStatus.PENDING) {
      return { success: false, message: 'Leave request is already processed' };
    }

    // 2. Lấy thông tin nhân viên
    const employee = await this.leaveRepo.getEmployeeById(leaveRequest.employeeId);
    if (!employee) {
      return { success: false, message: 'Employee not found' };
    }

    // 3. Kiểm tra Business Rule (Domain Service Level)
    const canApprove = this.leaveDomainService.canApproveLeave(employee, leaveRequest.days);
    if (!canApprove) {
      return { success: false, message: 'Insufficient leave balance' };
    }

    // 4. Lưu thay đổi (Infrastructure Level)
    await this.leaveRepo.updateLeaveStatus(leaveRequestId, LeaveStatus.APPROVED);
    await this.leaveRepo.updateEmployeeRemainingDays(employee.id, leaveRequest.days);

    return { success: true, message: 'Leave request approved successfully' };
  }
}
